features/FxA-71-AMO-migration-support/README.md

```diff
@@ -0,0 +1,189 @@
+across Mozilla's online properties.
+
+As an existing user of addons.mozilla.org,
+I want to clearly understand how and why
+I should migrate to a Firefox Account,
+so I can continue using the site
+with minimal interruption.
+
+
+## Measuring Success
+
+We will measure success of this feature
+by tracking whether users are successful
+at completing the migration.
+The key metric will be
+the number of users who successfully sign in or sign up,
+as a percentage of the number of users
+who enter the FxA login flow from AMO.
+
+
+## UX Mocks
+
+The overall UX for the migration flow
+was originally specified in https://github.com/mozilla/addons/pull/9.
+This feature captures the FxA-specific components
+so that we can properly manage and track their implementation.
+
+Three changes to our existing flows are required.
+
+### Migration indicator
+
+For users landing on the signup page
+from the AMO migration flow,
+we should show a highlight to inform them
+that they can use an existing account
+if they have one:
+
+![](migration-indicator.png)
+
+### Permission prompt
+
+Before redirecting back to AMO,
+we should prompt the user for permission
+to share their display-name and/or avatar
+if they have that information in their profile:
+
+![](permission-prompt.png)
+
+If the user has not set a display name or avatar,
+then that item should not appear in the prompt.
+
+### Smart signin/signup chooser
+
+AMO will provide the email address of an existing user
+when redirecting to Firefox Accounts
+for the migration flow.
+We should show the sign-in page
+when there is an existing account for that email,
+and the sign-up page when there is not.
+
+The provided email address
+should be pre-filled in the form
+in both cases.
+
+
+## Implementation details
+
+### Migration indicator
+
+In order to trigger the additional migration indicator,
+AMO should include `migration=amo` as a query parameter
+when beginning the OAuth login flow.
+
+In fxa-content-server,
+we will generalize the existing handling
+of the `migration` query parameter
+from the sync relier model,
+so that all reliers can import and report
+the source of a migration.
+We will also change the migration messaging
+from its current sync-specific text
+to the more generic version in the UX mocks above.
+
+### Permission prompt
+
+Since AMO is a trusted relier,
+it will not receive a permissions prompt by default.
+We will adopt the `prompt=consent` query parameter
+from [OpenID Connect](http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest)
+to allow them to request it explicitly.
+
+AMO should include `prompt=consent` as a query parameter
+when beginning the OAuth login flow.
+This will trigger fxa-content-server to show the permission prompt
+regardless of the trusted status of the relier.
+
+When showing the permission prompt
+the content-server should do the following:
+
+* Check and normalize the requested scope:
+  * If it is "profile" then replace it with
```

Comment:
AMO is not storing the OAuth tokens, they're completely transient.  The idea here is that AMO requests the broad "profile" scope, and we narrow it down to just the things we've prompted the user about.  AMO can then use those permissions and throw away the token, and we avoid the possibility of accidentally granting any additional profile:* scopes that we didn't prompt the user about.

If we add a new profile:* scope and the ability to prompt consent for it, subsequent AMO login requests would get that for free.

We could simplify by having AMO ask for "profile:uid profile:email profile:display_name profile:avatar" directly, but it allows for an accidental disconnect between the scopes we prompted for and the scopes that we actually deliver.